### PR TITLE
Mpuncel/fix races

### PR DESCRIPTION
**Pull request by mpuncel**

### CODE REVIEW

File: pkg/ds/daemon_set_test.go [hunk 1]
```diff
@@ -418,12 +444,37 @@ func TestPublishToReplication(t *testing.T) {
...
 	desiresErrCh := ds.WatchDesires(quitCh, updatedCh, deletedCh)
 	dsChangesErrCh := watchDSChanges(ds, dsStore, quitCh, updatedCh, deletedCh)
 
+	// We need to be careful when shutting down in order to avoid data
+	// races, since updatedCh and deletedCh are shared between two
+	// functions (WatchDesires() and watchDSChanges()). We need to make
+	// sure that we don't close updatedCh or deletedCh until their
+	// respective output channels have closed which indicates that they
+	// won't try to send any more values
+	defer func() {
```
Comment (mpuncel) [inline pkg/ds/daemon_set_test.go]: bonus: i think this is the cause of the `panic: send on closed channel` flake we have

**Comment by mpuncel** [comment]
it seems travis is catching even more races! so don't review until i have a green build :)

**Comment by mpuncel** [comment]
tests pass! this is ready for review

**Comment by petertseng** [comment]
Haven't yet looked at whether there is something doing this, but perhaps continuing to run the racy tests without the race detector?

File: pkg/replication/replication.go [hunk 2]
```diff
@@ -255,7 +259,7 @@ func (r *replication) Enact() {
 
 				go func() {
 					defer close(exitCh)
-					err = r.updateOne(node, timeoutCh, aggregateHealth)
```
Comment (rudle) [inline pkg/replication/replication.go]: I wonder if there's a linter we can turn on for "this goroutine captures free vars!" It's a dangerous property.

File: pkg/ds/daemon_set_test.go [hunk 3]
```diff
@@ -104,16 +104,17 @@ func watchDSChanges(
 				errCh <- util.Errorf("Error occured when watching daemon set changes: %v", watched.Err)
 			}
 
+			dsID := ds.ID()
```
Comment (rudle) [inline pkg/ds/daemon_set_test.go]: I'm curious: does this change fix a race? I can't figure out why it would.

Comment (mpuncel) [inline pkg/ds/daemon_set_test.go]: it doesn't fix a race, but theoretically the ID could change between `ID()` calls (although in practice it wouldn't).

I think this just came from me playing with the solution to this race for a while and one incarnation of it didn't involve locking inside of the `ID()` function

**Pull request merged**